ledger-core/test/integration/WalletFixture.hpp

```diff
@@ -0,0 +1,75 @@
+/*
+ *
+ * WalletFixture
+ * ledger-core
+ *
+ * Created by Alexis Le Provost on 27/01/2020.
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Ledger
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#pragma once
+
+#include <core/wallet/AbstractWallet.hpp>
+#include <core/wallet/WalletStore.hpp>
+#include <core/collections/DynamicObject.hpp>
+
+#include "BaseFixture.hpp"
+
+template <class WalletFactory>
+class WalletFixture : public BaseFixture {
+    using FunctionType = std::function<void(std::shared_ptr<AbstractWallet>, std::shared_ptr<WalletStore>)>;
+
+public:
+    enum class TestStrategy {
+        NEW_WALLET,
+        EXISTING_WALLET
+    };
+
+    void testWallet(
+        std::string const &name,
+        api::Currency const &currency,
+        TestStrategy strategy,
+        FunctionType const &f) {
+
+        auto services = newDefaultServices();
+        auto walletStore = newWalletStore(services);
+        auto factory = std::make_shared<WalletFactory>(currency, services);
+
+        wait(walletStore->addCurrency(currency)); 
+        walletStore->registerFactory(currency, factory);
+
+        auto wallet = [&]() {
+            switch (strategy) {
+                case TestStrategy::NEW_WALLET:
+                    return wait(walletStore->createWallet(name, currency.name, api::DynamicObject::newInstance()));
+                case TestStrategy::EXISTING_WALLET:
+                default:
+                    return wait(walletStore->getWallet(name));
+            }
+        }();
```

Comment:
This is called IIFE (for immediately invoked function expression). 

It allows complex variable initialization without the burden to declare an uninitialized variable (this type have a default constructor ? If yes, this default constructor may introduce some undesired side-effects).

Reply:
So as discussed out loud, this kind of trick is necessary because the language lacks some feature (the feature it lacks here is pattern-matching and statement-as-expressions).

However, you don’t need it either case. You could replace the whole thing with:

```cpp
auto wallet = strategy == TestTrategy::NEW_WALLET ?
  wait(walletStore->createWallet(name, currency.name, api::DynamicObject::newInstance())) :
  return wait(walletStore->getWallet(name));
```

Removing the visual noise.

Reply:
And just for the cultural note, in Rust, undefined variables are _safe_ until they’re used. So you could write:

```rust
let x;

match something {
  Foo => x = 1,
  Bar => x = 2,
}
```

Which compiles and is completly safe — but it’s so trivial we don’t do that. Instead, and about the feature C++ is missing, in Rust:

```rust
let x = match something {
  Foo => 1,
  Bar => 2,
};
```

Reply:
Thanks for the Rush example, really instructive. 

The technical choice to use IIFE here is to let the possibly to add new enumeration in future.
Since the need is not really fixed and probably will evolve with new coin integration; I prefer to avoid to use ternary statement here.